pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,844 @@
+func (s *ProfileSymbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*Location) {
+	for i, symLoc := range symLocs {
+		if len(symLoc.Lines) > 0 {
+			locIdx := locs[i].idx
+			profile.Location[locIdx].Line = make([]*googlev1.Line, len(symLoc.Lines))
+
+			for j, line := range symLoc.Lines {
+				// Create or find function name in string table
+				nameIdx := s.findOrAddString(profile, line.FunctionName)
+				filenameIdx := s.findOrAddString(profile, line.FilePath)
+
+				// Create or find function
+				funcId := s.findOrAddFunction(profile, nameIdx, filenameIdx, line.LineNumber)
+
+				profile.Location[locIdx].Line[j] = &googlev1.Line{
+					FunctionId: funcId,
+					Line:       int64(line.LineNumber),
+				}
+			}
+		}
+	}
+
+	mapping.HasFunctions = true
+	mapping.HasFilenames = true
+	mapping.HasLineNumbers = true
+}
+
+// findOrAddString finds a string in the string table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddString(profile *googlev1.Profile, str string) int64 {
+	for i, s := range profile.StringTable {
+		if s == str {
+			return int64(i)
+		}
+	}
+
+	idx := int64(len(profile.StringTable))
+	profile.StringTable = append(profile.StringTable, str)
+	return idx
+}
+
+// findOrAddFunction finds a function in the function table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddFunction(profile *googlev1.Profile, nameIdx, filenameIdx int64, lineNumber uint64) uint64 {
+	maxFuncId := uint64(0)
+	for _, f := range profile.Function {
+		maxFuncId = max(maxFuncId, f.Id)
+		if f.Name == nameIdx && f.Filename == filenameIdx {
+			return f.Id
+		}
+	}
+
+	funcId := maxFuncId + 1
+	profile.Function = append(profile.Function, &googlev1.Function{
+		Id:        funcId,
+		Name:      nameIdx,
+		Filename:  filenameIdx,
+		StartLine: int64(lineNumber),
+	})
+
+	return funcId
+}
+
+func (s *ProfileSymbolizer) Symbolize(ctx context.Context, req *Request) error {
+	start := time.Now()
+	status := StatusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	if s.checkSymbolCache(req) {
+		return nil
+	}
+
+	if s.checkLidiaTableCache(ctx, req) {
+		return nil
+	}
+
+	if s.checkObjectStoreCache(ctx, req) {
+		return nil
+	}
+
+	return s.fetchAndCacheFromDebuginfod(ctx, req, &status)
+}
+
+// symbolizeWithTable processes all locations in a request using a lidia table
+func (s *ProfileSymbolizer) symbolizeWithTable(_ context.Context, table *lidia.Table, ei *BinaryLayout, req *Request) error {
+	// Buffer for reusing in symbol lookups
+	var framesBuf []lidia.SourceInfoFrame
+
+	for _, loc := range req.Locations {
+		resolveStart := time.Now()
+		addr, err := MapRuntimeAddress(loc.Address, ei, Mapping{
+			Start:  loc.Mapping.Start,
+			Limit:  loc.Mapping.Limit,
+			Offset: loc.Mapping.Offset,
+		})
+		if err != nil {
+			return fmt.Errorf("normalize address: %w", err)
+		}
```

Comment:
I'd try to normalize all addresses before trying to resolve them.

Reply:
N/A as per https://github.com/grafana/pyroscope/pull/3799#discussion_r2063051024